ADR/ADR-003-PyBind11-For-CPP-Bindings.md

```diff
@@ -0,0 +1,419 @@
+* Declare a python module and its initialization (`PyInit_<mymodule>`)
+
+The code is built into a shared object (.so/.pyd) using the Python development
+headers and link flags. Python can then import this module directly. The module
+can be imported like any native Python package and the wrapped C++ functions
+can be called as normal Python functions.
+
+
+**Licence**: Python Software Foundation License (PSFL), BSD-style, GPL-compatible
+
+**Maturity**: Start in 1991 with python
+
+#### CFFI
+
+[source](https://github.com/python-cffi/cffi) |
+[documentation](https://cffi.readthedocs.io/en/stable/)
+
+CFFI (C Foreign Function Interface) is a Python library that allows Python code
+to call C functions and use C data types. With CFFI it is possible to
+ 
+* Extract C declarations from your to-be-wrapped API, note that this is a
+  preprocessing step because CFFI cannot parse C or C++ header files.
+* Load shared libraries using ffi.dlopen() and access all symbols. 
+* Call C functions and manipulate C data
+  structures as if they were native Python objects
+
+Applying CFFI yields a one-to-one mapping of your procedural code, to create an
+Object Oriented or even "Pythonic" interface you need to write another layer on
+top.
+
+Exposing complex mechanisms, for example an C++ iterator, requires writing and
+remapping intermediate C structures. The wrapping python code usually has to
+explicitly care about lifetime of objects. Python itself has no concept of
+ownership as it is considered in C/C++.
+
+CFFI does not require to distribute a native Python extension.
+
+**Licence**: MIT
+
+**Maturity**: Created in 2012, many releases (33 tags on github), good documentation. 
+
+**Longevity**: 8 maintainers (pypi), 77 contributors (GitHub).
+
+**Activity**: Multiple releases in the last years, community interaction.
+
+**Visibility**: 204 stars on GitHub.
+
+#### PyBind11 3.0
+
+[source](https://github.com/pybind/pybind11) |
+[documentation](https://pybind11.readthedocs.io/en/stable/)
+
+PyBind11 is a C++ header-only library that enables seamless binding of C++ code
+to Python. It allows you to expose C++ classes, functions, and data structures
+to Python with minimal boilerplate. 
+
+Pybind11 features:
+
+* Automatically translates C++ exceptions to matching Python exceptions, see
+  [details](https://pybind11.readthedocs.io/en/stable/advanced/exceptions.html).
+  This includes the ability to register custom type conversions that map domain specific
+  exceptions, e.g. `eckit::Exception` to `eckit.Excpeption`.
+* Automatically maps between many std types and corresponding python types, see
+  [here](https://pybind11.readthedocs.io/en/stable/advanced/cast/overview.html#conversion-table).
+* Allows exchanging wrapped types across bindings, see [module local
+  bindings](https://pybind11.readthedocs.io/en/stable/advanced/classes.html#module-local-class-bindings)
+* Allows setting call policies that affect object lifetimes. These policies allow to
```

Comment:
"These policies enable the definition of dependencies ..." or similar. "allow to" is bad English.